src/uffd_handler/src/handler.rs

```diff
@@ -65,18 +67,31 @@ impl UffdManager for Uffd {
     fn populate_from_file(
         &self,
         buff: *const libc::c_void,
+        start_addr: usize,
+        end_addr: usize,
         region: &MemRegion,
     ) -> Result<(usize, usize)> {
-        let src = buff as u64 + region.mapping.offset;
-        let start_addr = region.mapping.base_host_virt_addr;
-        let len = region.mapping.size;
-        // Populate whole region from backing mem-file.
-        // This offers an example of how memory can be loaded in RAM,
-        // however this can be adjusted to accommodate use case needs.
+        let page_size = get_page_size().map_err(HandlerError::PageSize)?;
+        // The result will always be positive because both start_addr and end_addr are within the
+        // current region and the difference between them is always at least page size.
+        let mut len = end_addr - start_addr;
+        // Length to uffd copy must be multiple of page size.
+        if len % page_size != 0 {
+            len = page_size * (len / page_size + 1);
```

Comment:
> we can avoid multiplication/division via len += page_size - remainder

You are right! I have no idea why I overly complicated this check, haha. 

> And since page_size is always a multiple of 2, we can even do something like len = (len | (page_size - 1)) + 1, but that's very obscure and probably shouldnt be done unless we really care about every last ns of performance here, haha.

I would go with your first suggestion, I agree that `len = (len | (page_size - 1)) + 1` looks obscure and it's harder to digest.